### PR TITLE
Release 2.0: Major refactor

### PR DESCRIPTION
**Pull request by tobilarscheid**
Hi,

this is the consolidation of our Pull Requests (#5, #4, #3).

As described, App Role is used as the only mechanism to authenticate against vault. Role Id and Secret are stored in a dedicated Credential Type.

We added support for the Folders Plugin, so you can configure the Plugin on a folder level (in addition to global and on job level) and assign folder-scoped credentials.

We also completely rewrote the tests, they do no longer rely on a running Vault instance. (How? We wrapped all Vault calls in a `VaultAccessor` which we then mock during tests) This will hopefully make contributing a lot more convenient for others.

### CODE REVIEW

**Comment by schmitzhermes** [comment]
Moreover we should mention that we rewrote the entire configuration mechanism. The config is not longer tightly coupled to the `VaultBuildWrapper` but is generalized in the `VaultConfiguration`. The `VaultConfigResolver` determines the config that should be used in the current setting.

**Comment by tobilarscheid** [comment]
@ptierno see the updated readme for a better description of how the plugin would work now.

**Comment by schmitzhermes** [comment]
the failing test is going to be fixed tomorrow. There seems to be an issue when masking the credentials. This issue is only experienced in the integration test, but not in a real Jenkins instance. We are going to have a look at it.

**Comment by ptierno** [comment]
@tobilarscheid @schmitzhermes I appreciate the work you have put into this and will be working on getting 2.0 released shortly (assuming tests are passing). The one thing I want to do before moving on with 2.0 is to reopen #5 and to allow the ability to read tokens from a file as suggested in [JENKINS-37713](issues.jenkins-ci.org/browse/JENKINS-37713). I would like users to be able to pin to a 1.4 release if for some reason they dont want to use AppRole or if they are using an older version of vault which still used AppId. Let me know your thoughts.

**Comment by jantman** [comment]
For a bit of user feedback, it looks like we'll pin to 1.4 if this comes out, as we will never use AppRole (we view it as horribly insecure). I strongly feel that *how* a token gets into Jenkins should be up to the user, and that making this plugin specific to AppRole isn't in the best interest of most users, who as far as I know, don't use that.

We use the EC2 Auth backend via Hashicorp's VSI (Vault Secure Introduction client, an Enterprise/paid product), and a "sidecar" daemon we wrote that generates a child token with a 15 minute TTL and provisions that into Jenkins. Until now we've been provisioning via a custom job that sets the token in the plugin config and gets triggered by our sidecar app, but now that 1.4 is out, we'll be writing the token to the file, and possibly dropping the TTL from 15 minutes to something more secure like 60 seconds.

**Comment by schmitzhermes** [comment]
Hey @jantman, thanks for your comment. First of all I'd like to know: why exactly do you consider AppRole as "horribly insecure"? Why is generating a Token inside Jenkins (in exchange for `role_id` and `secret_id`) more "insecure" than just writing a custom algorithm that puts a token inside Jenkins? Especially if you can tell Vault only to accept `role_id` and `secret_id` from a specific CIDR range in exchange for a token. 

However, I do agree with you, that it may not be in the best interest of _all_ users to couple this plugin tightly to the AppRole mechanism. I somehow feel, though, that reading a token from the filesystem feels "hacky", but I might be wrong. What would be your preferred way to retrieve the token?

**Comment by jantman** [comment]
@schmitzhermes So first of all, our overriding principle is that tokens should be short-lived (which seems to be what HashiCorp pushes in their training and docs). We're aiming for tokens that are rolled every 1-5 minutes.

Ok, maybe "horribly insecure" was rather hyperbolic. However in our view, and I think the view that Hashicorp would like to promulgate, tokens should be generated by trusted third parties. The EC2 Authentication backend allows vault to reliably identify a single EC2 instance (based on a variety of factors including AMI, IAM Role, etc.). It's essentially a first-secret-less authentication mechanism; it relies on AWS to be able to identify that a request is coming from where it says. There's a PR in the pipeline now to generalize that for anything on AWS that has an IAM role. In the container world, Hashicorp seems to strongly suggest that token generation should be up to your job scheduler; that Mesos or k18n or whatever else should generate a token for a container, and push it into said container.

My/our standpoint on AppRole is essentially that (1) anything which requires a "secret" being provisioned into an end instance is a problem, since that becomes the First Secret problem all over again; we want hands-off automated management, without hard-coding part of the secret, and, (2) all of the suggestions we've seen for ``secret_id`` either require manual intervention, or can be trivially discovered by anyone with access to our environment.

Also I imagine we're not alone in saying that limiting by CIDR is useless to us; we're running primarily in AWS, and have a number of *giant* subnets that are shared between multiple teams, so CIDR is virtually meaningless.

My casual involvement with the Vault community and development has given me the impression that there's substantial bias in favor of authentication mechanisms that either use some trusted third party to establish authentication (like the EC2 Auth backend), or that use a trusted third party to *push* tokens into the app environment.

Looking at the number [authentication backends](https://www.vaultproject.io/docs/auth/index.html) that Vault currently supports, and how quickly ones have been added lately, makes me very much feel like it's problematic in the long term to tie vault clients like this plugin to a specific auth method, and that it's better to leave authentication itself to some intermediary.

To me, reading the token from the filesystem seems to be the right way. It's also the only way currently supported by hashicorp's Vault Secure Introduction (VSI) client, so I'd say that's ample proof on their end that it's a suggested method. Our current plan is to run VSI to authenticate our EC2 instances and keep their tokens renewed. We then run a daemon that we wrote internally which just uses the VSI "instance" token to generate a 60-second-TTL child token (every 60 seconds) and write that to a file. As such, we've auth'ed based on our EC2 instance (no "first secret" at all, the instance just needs to know the vault address, that's it) and Jenkins itself only has access to a token with a 60s TTL, so even if it's leaked, there's a very small window to exploit it.

**Comment by ptierno** [comment]
@schmitzhermes @tobilarscheid would you be willing to implement this while still keeping the current functionality? i dont want to lose your work but i agree with @jantman.. maybe have a vault role credential a vault token credential and a vault token file credential and the plugin can auth differently depending on what is provided?

**Comment by schmitzhermes** [comment]
Hey @ptierno, hey @jantman,
I agree that limiting this plugin to only AppRole might be problematic. We might want to implement storing the token on disk instead of using AppRole. In fact, @tobilarscheid already implemented this one in cc33ad06b15fb88e1f3753dfb9d0f2c437259199 of #5.
Adapting this to 2.0 might take some time though, since there are a couple of design decisions to be made. I am going to talk to Tobi about this.

**Comment by ptierno** [comment]
@schmitzhermes thank you, let me know what you two decide and how you want to move forward with implementing this.

**Comment by tobilarscheid** [comment]
@jantman , @ptierno 

thanks for your detailed and constructive feedback. I really enjoy the exchange!

I have restored the token based authentication, either with a token stored in Jenkins or in a File.

The authentication mechanism is now more generic than ever, the configured credential type essentially does the whole authentication - implementing additional ones should be very easy and clear from now. Refer to the documentation for details.

Also I want to point out how helpful the numerous tests were for this refactoring - this is definitely a thing we should adhere to.

**Comment by ptierno** [comment]
@tobilarscheid @schmitzhermes ill be working on getting this merged and released as 1.5 soon. Thanks.

**Comment by schmitzhermes** [comment]
@ptierno Great. But this is a major version change imho - I'd go for 2.0.

**Comment by ptierno** [comment]
@schmitzhermes thats fine. we will consider it a 2.o release

**Comment by tobilarscheid** [comment]
Great, thanks for your effort @ptierno !

**Comment by mattva01** [comment]
This is awesome! There is one thing I'd like to see as an option at some point, and that's the ability to use a response-wrapped value (probably/potentially from a file) that the plugin can unwrap to get the actual secret-id. It just feels kinda odd to have a human handling both the app-role and secret directly, though obviously your Jenkins admin should be a trusted individual anyway.

**Comment by schmitzhermes** [comment]
That's a cool idea. Let's consider this for 2.1 since this PR is already pretty bloated. :-)

**Comment by tobilarscheid** [comment]
Sounds cool, @mattva01 can you open an issue for that so we remember it? I agree with @schmitzhermes that this is probably too much for 2.0!

**Comment by mattva01** [comment]
Sure, give me a bit.  I certainly agree that it shouldn't hold up 2.0!  I mainly put it here because it was where the AppRole discussion was happening.

File: README.md [hunk 1]
```diff
@@ -1,61 +1,107 @@
-# jenkins-vault-plugin
+[![Build Status](https://jenkins.ci.cloudbees.com/buildStatus/icon?job=plugins/hashicorp-vault-plugin&build=20)](https://jenkins.ci.cloudbees.com/job/plugins/hashicorp-vault-plugin/20/)
+# Jenkins Vault Plugin
 
-This plugin adds a build wrapper to set environment variables from a HashiCorp [Vault](https://www.vaultproject.io/) secret.
+This plugin adds a build wrapper to set environment variables from a HashiCorp [Vault](https://www.vaultproject.io/) secret. Secrets are generally masked in the build log, so you can't accidentally print them.
 
-## Caveats
+# Vault Authentication Backends
```
Comment (ptierno) [inline README.md]: Update this section to account for file token credential and token credential

File: README.md [hunk 2]
```diff
@@ -1,61 +1,107 @@
...
+### Upgrade from 1.x to 2.0
+The `BuildWrapper` did not change, so no changes to your Jenkinsfile should be necessary. However, you need to reconfigure Vault in your Jenkins instance based on the instructions above. There is no way to smoothly upgrade this, because this is a major rewrite and handling of configuration completly changed.
+# CHANGELOG
 
-* **2017/04/10** - Feature Release - 1.4
-  * Support reading Vault Token from file on disk [JENKINS-37713](issues.jenkins-ci.org/browse/JENKINS-37713)
-  * Using credentials plugin for authentication token [JENKINS-38646](issues.jenkins-ci.org/browse/JENKINS-38646)
-* **2017/03/03** - Feature Release - 1.3
-  * Vault Plugin should mask credentials in build log [JENKINS-39383](issues.jenkins-ci.org/browse/JENKINS-39383)
+* **2017/05/04** - Breaking change release (AppRole auth backend, Folder ability, improved configuration, ...)
```
Comment (ptierno) [inline README.md]: If you can align this with the current master branches changelog that would be great

File: README.md (hunk 1 above)
Comment (tobilarscheid) [inline README.md]: This should be in line 8? Or what else do you mean?
> Furthermore, this plugin allows using a Vault Token - either configured directly in Jenkins or read from an arbitrary file on the Jenkins Machine.

File: README.md [hunk 3]
```diff
@@ -1,61 +1,112 @@
-# jenkins-vault-plugin
+[![Build Status](https://jenkins.ci.cloudbees.com/buildStatus/icon?job=plugins/hashicorp-vault-plugin&build=20)](https://jenkins.ci.cloudbees.com/job/plugins/hashicorp-vault-plugin/20/)
+# Jenkins Vault Plugin
 
-This plugin adds a build wrapper to set environment variables from a HashiCorp [Vault](https://www.vaultproject.io/) secret.
+This plugin adds a build wrapper to set environment variables from a HashiCorp [Vault](https://www.vaultproject.io/) secret. Secrets are generally masked in the build log, so you can't accidentally print them.
 
-## Caveats
+# Vault Authentication Backends
+This plugin only allows authenticating against Vault using the AppRole authentication backend. Hashicorp recommends using AppRole for Servers / automated workflows (like Jenkins) and using Tokens (default mechanism, Github Token, ...) for every developer's machine.
```
Comment (ptierno) [inline README.md]: I mean line 7 where it states that the plugin only allows authenticating with approle.

Comment (tobilarscheid) [inline README.md]: sorry... my bad... I'll fix it

**Pull request merged**